File: 2015/php/1/main.php
<?php

function main() : void {
    // get command line arguments
    $args = $_SERVER['argv'];
    // remove the first argument which is the script name
    array_shift($args);
    // get the first argument
    if(count($args) < 1) {
        echo "Usage: php main.php <input_file>\n";
        return;
    }

    $input_file = $args[0];

    // read the input file
    $input = file_get_contents($input_file);

    // loop each line
    $lines = explode("\n", $input);
    // loop each character
    $floor = 0;
    $count = 0;
    $basement = 0;
    foreach($lines as $line) {
        $chars = str_split($line);
        foreach($chars as $char) {
            if($char == '(') {
                $floor++;
            }
            elseif($char == ')') {
                $floor--;
            }

            $count++;
            if($floor == -1 && $basement == 0) {
                $basement = $count;
            }
        }
    }

    echo "Part 1: " . $floor . "\n";
    echo "Part 2: " . $basement . "\n";
}

main();

File: 2015/rust/1/src/main.rs
use std::env;
fn main() {
    // get command line arguments
    let args: Vec<String> = env::args().collect();

    if args.len() < 2 {
        println!("Usage: {} <input_file>", args[0]);
        return;
    }

    let input_file = &args[1];

    // read the file
    let contents =
        std::fs::read_to_string(input_file).expect("Something went wrong reading the file");

    let mut floor = 0;
    let mut count = 0;
    let mut basement = 0;
    // loop through the lines
    for line in contents.lines() {
        // loop through the characters
        for c in line.chars() {
            if c == '(' {
                floor += 1;
            } else if c == ')' {
                floor -= 1;
            }

            count += 1;
            if floor == -1 && basement == 0 {
                basement = count;
            }
        }
    }
    println!("Part 1: {}", floor);
    println!("Part 2: {}", basement);
}
